File: src/main/java/com/soapdogg/lispInterpreter/functions/DefunFunction.java
package com.soapdogg.lispInterpreter.functions;

import com.soapdogg.lispInterpreter.datamodels.ExpressionNode;
import com.soapdogg.lispInterpreter.datamodels.Node;
import com.soapdogg.lispInterpreter.datamodels.UserDefinedFunction;
import com.soapdogg.lispInterpreter.generator.UserDefinedFunctionFormalParameterGenerator;
import com.soapdogg.lispInterpreter.valueretriver.AtomicValueRetriever;
import com.soapdogg.lispInterpreter.valueretriver.ListValueRetriever;
import com.soapdogg.lispInterpreter.asserter.UserDefinedFormalParametersAsserter;
import com.soapdogg.lispInterpreter.asserter.UserDefinedFunctionNameAsserter;
import com.soapdogg.lispInterpreter.constants.FunctionLengthConstants;
import com.soapdogg.lispInterpreter.constants.FunctionNameConstants;
import com.soapdogg.lispInterpreter.asserter.FunctionLengthAsserter;
import lombok.AllArgsConstructor;

import java.util.*;

@AllArgsConstructor(staticName = "newInstance")
public class DefunFunction  {

    private final FunctionLengthAsserter functionLengthAsserter;
    private final AtomicValueRetriever atomicValueRetriever;
    private final ListValueRetriever listValueRetriever;
    private final UserDefinedFunctionNameAsserter userDefinedFunctionNameAsserter;
    private final UserDefinedFunctionFormalParameterGenerator userDefinedFunctionFormalParameterGenerator;
    private final UserDefinedFormalParametersAsserter userDefinedFormalParametersAsserter;

    public UserDefinedFunction evaluateLispFunction(
        final Node params
    ) throws Exception {
        functionLengthAsserter.assertLengthIsAsExpected(
            FunctionNameConstants.DEFUN,
            FunctionLengthConstants.FOUR,
            params
        );

        ExpressionNode functionNameNode = listValueRetriever.retrieveListValue(
            params,
            FunctionNameConstants.DEFUN,
            Collections.emptyMap()
        );
        String functionName = atomicValueRetriever.retrieveAtomicValue(
            functionNameNode.getAddress(),
            FunctionLengthConstants.ONE,
            FunctionNameConstants.DEFUN
        );
        userDefinedFunctionNameAsserter.assertFunctionNameIsValid(functionName);

        Node functionNameNodeData = functionNameNode.getData();
        ExpressionNode tempNode = listValueRetriever.retrieveListValue(
            functionNameNodeData,
            FunctionNameConstants.DEFUN,
            Collections.emptyMap()
        );
        List<String> formalParameters = userDefinedFunctionFormalParameterGenerator.getFormalParameters(
            tempNode.getAddress(),
            FunctionLengthConstants.ONE,
            Collections.emptyMap()
        );
        userDefinedFormalParametersAsserter.assertFormalParameters(formalParameters);

        Node body = tempNode.getData();

        return new UserDefinedFunction(
            formalParameters,
            body,
            functionName
        );
    }
}

File: src/main/java/com/soapdogg/lispInterpreter/functions/DefunFunction.kt
package com.soapdogg.lispInterpreter.functions

import com.soapdogg.lispInterpreter.asserter.FunctionLengthAsserter
import com.soapdogg.lispInterpreter.asserter.UserDefinedFormalParametersAsserter
import com.soapdogg.lispInterpreter.asserter.UserDefinedFunctionNameAsserter
import com.soapdogg.lispInterpreter.constants.FunctionLengthConstants
import com.soapdogg.lispInterpreter.constants.FunctionNameConstants
import com.soapdogg.lispInterpreter.converter.NodeConverter
import com.soapdogg.lispInterpreter.datamodels.AtomNode
import com.soapdogg.lispInterpreter.datamodels.ExpressionListNode
import com.soapdogg.lispInterpreter.datamodels.UserDefinedFunction
import com.soapdogg.lispInterpreter.generator.UserDefinedFunctionFormalParameterGenerator

class DefunFunction(
    private val functionLengthAsserter: FunctionLengthAsserter,
    private val userDefinedFunctionNameAsserter: UserDefinedFunctionNameAsserter,
    private val userDefinedFunctionFormalParameterGenerator: UserDefinedFunctionFormalParameterGenerator,
    private val userDefinedFormalParametersAsserter: UserDefinedFormalParametersAsserter,
    private val nodeConverter: NodeConverter
) {

    fun evaluateLispFunction(
        params: ExpressionListNode
    ): UserDefinedFunction {
        functionLengthAsserter.assertLengthIsAsExpected(
            FunctionNameConstants.DEFUN,
            FunctionLengthConstants.FOUR,
            params
        )
        val functionNameNode = params.children[1] as AtomNode
        val functionName = functionNameNode.value
        userDefinedFunctionNameAsserter.assertFunctionNameIsValid(functionName)

        val formalParameters = userDefinedFunctionFormalParameterGenerator.getFormalParameters(
            nodeConverter.convertNodeV2ToNode(params.children[2]),
            FunctionLengthConstants.ONE,
            mapOf()
        )
        userDefinedFormalParametersAsserter.assertFormalParameters(formalParameters)
        val convertedBody = nodeConverter.convertNodeV2ToNode(params.children[3])
        return UserDefinedFunction(
            formalParameters,
            convertedBody,
            functionName
        )
    }
}